rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/locations/impl/TreeSearch.java

```diff
@@ -0,0 +1,124 @@
+     * @param line    line to check if its within the range
+     * @param column  column to see if we are within the range
+     * @return true iff the given `loc` range spans around the line/column position, inclusively.
+     */
+    private static boolean inside(ISourceLocation loc, int line, int column) {
+        if (!loc.hasLineColumn()) {
+            return false;
+        }
+
+        if (line < loc.getBeginLine() || line > loc.getEndLine()) {
+            return false;
+        }
+
+        if (line == loc.getBeginLine()) {
+            // then make sure we are not before the start
+            return loc.getBeginColumn() <= column;
+        }
+
+        if (line == loc.getEndLine()) {
+            // then make sure we are not beyond the end
+            return column <= loc.getEndColumn();
+        }
+
+        // otherwise we are in the middle of the lines 
+        // and the column is irrelevant
+
+        return true;
+    }
+    
+    /**
+     * Produces a list of trees that are "in focus" at given line and column offset (UTF-24).
+     * 
+     * This log(filesize) algorithm quickly collects the trees along a spine from the
+     * root to the smallest lexical or context-free node. The list is returned in
+     * reverse order such that you can select the "most specific" tree by starting
+     * at the start of the list.
+     * 
+     * @param tree   parse tree
+     * @param line   search term in lines (1-based)
+     * @param column search term in columns (0-based, UTF-32 codepoints)
+     * @return list of tree that are around the given line/column position, ordered from child to parent.
+     */
+    public static IList computeFocusList(ITree tree, int line, int column) {
+        var lw = IRascalValueFactory.getInstance().listWriter();
+        computeFocusList(lw, tree, line, column);
+        return lw.done();
+    }
+
+    private static boolean computeFocusList(IListWriter focus, ITree tree, int line, int column) {
+		ISourceLocation l = TreeAdapter.getLocation(tree);
+
+		if (l == null) {
+            // inside a layout, literal or character
+			return false;
+		}
+
+		if (TreeAdapter.isLexical(tree)) {
+			if (inside(l, line, column)) {
+                focus.append(tree);
+                // stop and return success
+				return true;
+			}   
+            else {
+                // stop and return failure
+                return false;
+            }
+		}
+
+		if (TreeAdapter.isAmb(tree) && !tree.getAlternatives().isEmpty()) {
+            // pick any tree to make the best of it.
+			return computeFocusList(focus, (ITree) tree.getAlternatives().iterator().next(), line, column);
+		}
+
+		if (TreeAdapter.isAppl(tree)) {
+			IList children = TreeAdapter.getASTArgs(tree); // this skips layout trees
+
+			for (IValue child : children) {
+				ISourceLocation childLoc = TreeAdapter.getLocation((ITree) child);
+
+				if (childLoc == null) {
+					continue;
+				}
+
+				if (inside(childLoc, line, column)) {
```

Comment:
this logic seems the same as the start of this function, shouldn't we let recursion take care to only look at the interesting childs? and just only check the result of the `computeFocusList`?

Reply:
No it's different. We check one time extra that the literal is inside, but that is really a superfluous assert for documentation purposes. Once you arrive at the lexical level, you're there because that's the onl way you got there via line 106. The exception is of course if someone provides only a toplevel lexical tree, and this extra tests ensures sanity in such a case. 

We only recurse for context-free children which are `inside` on line 106.